web-api/src/middleware/request_context.rs

```diff
@@ -0,0 +1,184 @@
+    /// Tries to return the current `ApiCallContext` based on a request.
+    ///
+    /// The context will be setup by the `setup_request_context` middleware.
+    #[instrument(skip_all, err)]
+    pub(crate) fn try_extract_from_request<R>(
+        request: &HttpRequest,
+        func: impl FnOnce(&Arc<RequestContext>) -> R,
+    ) -> Result<R, AccessError> {
+        let extensions = request.extensions();
+        Ok(func(extensions.get::<Arc<RequestContext>>().ok_or(
+            AccessError {
+                method: "try_extract_from_request",
+            },
+        )?))
+    }
+}
+
+#[derive(Debug, Error)]
+#[error("Failed to access expected context value in: {method}")]
+pub(crate) struct AccessError {
+    method: &'static str,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display, PartialEq, Deserialize, Serialize)]
+#[serde(transparent)]
+pub(crate) struct TenantId(Uuid);
+
+impl TenantId {
+    pub(crate) fn missing() -> Self {
+        TenantId(Uuid::nil())
+    }
+}
+
+impl TryFrom<&'_ HeaderValue> for TenantId {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &HeaderValue) -> Result<Self, Self::Error> {
+        let header = value.to_str().map_err(|_| {
+            anyhow!(
+                "Non Utf-8 compatible {TENANT_ID_HEADER} header: {:?}",
+                String::from_utf8_lossy(value.as_bytes())
+            )
+        })?;
+        header
+            .parse::<Uuid>()
+            .map_err(|_| anyhow!("Non UUID {TENANT_ID_HEADER} header: {header}"))
+            .map(TenantId)
+    }
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display, Serialize)]
+#[serde(transparent)]
+pub(crate) struct RequestId(Uuid);
+
+impl RequestId {
+    pub(crate) fn generate() -> Self {
+        Self(Uuid::new_v4())
+    }
+
+    pub(crate) const fn missing() -> Self {
+        Self(Uuid::nil())
+    }
+}
+
+/// Setups the call context.
+///
+/// This make the `RequestId` and `TenantId` available as extensions and sets up tracing for all calls.
+///
+/// The `TenantId` is seen as _required_, i.e. the middleware will fail any
+/// request where it can't find a tenant id.
+pub(crate) fn setup_request_context<A, S>(
+    request: ServiceRequest,
+    service: &S,
+) -> impl Future<Output = Result<ServiceResponse<BoxBody>, actix_web::Error>> + 'static
+where
+    A: AsRef<tenants::Config> + 'static,
+    S: Service<ServiceRequest, Response = ServiceResponse<BoxBody>, Error = actix_web::Error>,
+    S::Future: 'static,
+{
+    let request_id = RequestId::generate();
+    let config = request.app_data::<Data<A>>().unwrap().get_ref().as_ref();
+
+    let tenant_id = match extract_tenant_id(config, &request) {
+        Ok(id) => id,
+        Err(error) => {
+            let response = middleware_failure(
+                "setup_request_context",
+                request,
+                Some(request_id),
+                None,
+                error,
+            );
+            return Either::Left(future::ok(response));
```

Comment:
We should not spend any though on weather we use Right or Left in a context where we return a `impl Trait`.

The whole idea of `Either` compared to `Result`/a custom enum  is that while both sides have different types they have the same kind of API and are API wise roughly equivalent.  Anything beyond  is an implementation detail (e.g. the left future always resolves to an error). At least in rust where very intentionally types like `Option` and `Result` where added instead of having a more generic `Either`. 

Convention wise using `Left` first then `Right` is the only common convention. Which is also why e.g. the `either` library maps `Result<R,L> to Either<L, R>` as in proper (rust) usages of `Either` the the error case most times happen as a form of "early return" and in turn befor the normal result. I.e. it happens in a place people tend to use `Left` for.

But anyway this discussion is kinda pointless as the `Either` parameter order isn't part of the API due to `impl Trait`.